File: hash/programmers_42579/solution.py
def solution(genres, plays):
    answer = []
    total_plays = {}
    genre_songs = {}
    
    for i in range(len(genres)):
        genre = genres[i]
        play = plays[i]
        song_id = i
        
        total_plays[genre] = total_plays.get(genre, 0) + play
        
        if genre not in genre_songs:
            genre_songs[genre] = []

        genre_songs[genre].append({'id': song_id, 'play': play})
        
    sorted_total_plays = sorted(total_plays.items(), key=lambda item: item[1], reverse=True)
    
    for genre, total_play in sorted_total_plays:
        songs_with_genre = sorted(genre_songs[genre], key=lambda item: (-item['play'], item['id']))
        
        for song in songs_with_genre[:2]: 
            answer.append(song['id'])
    
    return answer

File: hash/programmers_42579/solution.js
function solution(genres, plays) {
    let answer = [];
    const totalPlays = new Map();
    const genrePlays = new Map();
    
    for (let i = 0; i < genres.length; i++) {
        const genre = genres[i];
        const play = plays[i];
        const id = i;

        totalPlays.set(genre, (totalPlays.get(genre) || 0) + play);
        
        const songsInCurrentGenre = genrePlays.get(genre) || [];
        
        songsInCurrentGenre.push({ id: id, play: play });
        
        genrePlays.set(genre, songsInCurrentGenre);
    }
    
    const sortedTotalPlays = [...totalPlays.entries()].sort(([, a], [, b]) => b - a);
    
    for (const [genre, totalPlay] of sortedTotalPlays) {
        const songsWithGenre = genrePlays.get(genre).sort((a, b) => {
            if (a.play === b.play) {
                return a.id - b.id;
            }
            return b.play - a.play;
        });

        for (let i = 0; i < Math.min(songsWithGenre.length, 2); i++) {
            answer.push(songsWithGenre[i].id);
        }
    }

    return answer;
}